### ISSUE TITLE
Not all BIOSes are guaranteed to jump to 0x0000:0x7C00

### ISSUE THREAD

**Issue by cntkillme** [JustAPerson/vos#1]
Some BIOSes jump to `0x07C0:0x0000` while others to `0x0000:0x7C00` which both map to the same physical address of `0x07C00`. The rule of thumb is that assume everything is wrong except the physical address and the contents of `DL`.

This can affect the bootloader because if the BIOS did, in fact, jump to `0x07C0:0x0000` then setting `ds` to 0 may cause problems. Also when you do: `mov [drive_number], dl` before ever initializing `ds`, it should be assumed that it could cause problems because the `ds` is not guaranteed to be set to `0x07C0` meaning it might not be mapped to `0x0000:offsetFrom7C00`.

**Comment by JustAPerson** [comment]
Sorry for the delay. I've finally gotten back into this, and I'll be pushing some commits to update things and simplify the build process this weekend. Anyway, the current intention is only to support qemu's hardware for the time being since that's really the only thing I can test on. However, since the bootloaders have plenty of spare space, I wouldn't mind addressing this.

However, if I'm understanding this correctly, wouldn't it be `CS` set to `0x07c0` in that case? Seems like this could be resolved with a far jump to clear `CS`. Though you're right, we should clear `DS` sooner.

**Comment by cntkillme** [comment]
Yeah. What I might do is a far jump to 0x07C0:stage1 at the top then you would only have to worry about setting DS to 0x07C0. I'd imagine something like this should be enough:

```
jmp 0x07C0:stage1
stage1:
mov ax, cs
mov ds, ax
```

**Comment by JustAPerson** [comment]
Hmm, I'm confused. Is this in reference to the [system bootloader](https://github.com/JustAPerson/vos/blob/master/src/boot/alpha.s) or [volume bootloader](https://github.com/JustAPerson/vos/blob/master/src/boot/beta.s)?

With the former, I think it's easier to set `cs`to `0x0000` with a far jump and clear `ds` early since the system bootloader relocates immediately.

**Comment by cntkillme** [comment]
In volume bootloader: you're indexing memory when the DS is not yet initialized.
**Edit** I believe you're doing it in both (indexing memory before initializing DS) as the DS segment is implied when you index memory in real mode (although you can override the segment technically).
